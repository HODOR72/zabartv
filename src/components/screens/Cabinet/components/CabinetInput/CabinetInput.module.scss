.item {
	position: relative;
	border: 1px solid #232c3d;
	border-radius: 25px;
	padding: 32px 45px 37px;
	transition: border-color 0.3s;

	@media (max-width: 1200px) {
		padding: 27px 30px;
	}

	@media (max-width: 768px) {
		padding: 32px 45px 37px;
	}

	@media (max-width: 360px) {
		padding: 32px 35px 30px;
	}
}

[data-theme='light'] {
	.item {
		border-color: #dadee7;
	}
}

.label {
	display: block;
	font-weight: 500;
	font-size: 16px;
	line-height: 27.8px;
	color: #52596c;
}

.input {
	-webkit-appearance: none;
	width: 100%;
	appearance: none;
	padding: 0;
	border: none;
	border-radius: 0;
	background-color: transparent;
	outline: none;
	font-family: inherit;
	color: var(--color-light);
	transition: color 0.3s;
	font-size: 20px;
	line-height: 33px;

	&::placeholder {
		opacity: 1;
		color: var(--color-light);
	}
}

.value {
	pointer-events: none;
	cursor: default;
	margin-top: 0;

	// &[type='password'] {
	// 	font-family: Verdana;
	// 	letter-spacing: 2.2px;
	// }
}

.apply {
	margin-top: 5px;
}

.btn {
	position: absolute;
	display: block;
	right: 20px;
	top: 20px;
	width: 22px;
	height: 22px;

	svg {
		display: block;
		width: 100%;
		height: 100%;
		fill: var(--color-light);
	}
}
