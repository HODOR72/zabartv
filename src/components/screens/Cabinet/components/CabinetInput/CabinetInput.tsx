import { ChangeEvent, FC, InputHTMLAttributes, useRef, useState } from 'react';
import { ButtonBase } from '@/components/ButtonBase/ButtonBase';
import { EditIcon, CloseIcon } from '@/icons';
import { Link } from '@/UI/Link/Link';
import classNames from 'classnames';
import styles from './CabinetInput.module.scss';

export interface CabinetInputProps extends InputHTMLAttributes<HTMLInputElement> {
	label: string;
	date?: boolean;
	value: string;
	applyChanges: (value: string) => void;
}

export const CabinetInput: FC<CabinetInputProps> = ({
	className,
	applyChanges,
	type,
	value,
	date,
	label,
	...props
}) => {
	const inputRef = useRef<HTMLInputElement>(null);
	const [isEdited, setIsEdited] = useState<boolean>(false);
	const [inputValue, setInputValue] = useState<string>(value);

	const closeField = () => {
		inputRef.current?.blur();
		setIsEdited(false);
	};

	const handleEditInput = () => {
		if (!isEdited) {
			setIsEdited(true);
			inputRef.current?.focus();
		} else {
			setInputValue(value);
			closeField();
		}
	};

	const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
		setInputValue(e.target.value);
	};

	const handleApplyChanges = () => {
		applyChanges(inputValue);
		closeField();
	};

	return (
		<div data-testid="cabinet-input-item" className={styles.item}>
			<span className={styles.label}>{label}</span>
			<input
				data-testid="cabinet-input"
				ref={inputRef}
				className={classNames(styles.input, !isEdited && styles.value)}
				type={isEdited ? 'text' : type}
				placeholder={value}
				value={inputValue}
				onChange={handleChange}
				{...props}
			/>
			{isEdited && (
				<Link
					disabled={inputValue.length === 0}
					as="button"
					onClick={handleApplyChanges}
					className={styles.apply}
				>
					Сохранить изменения
				</Link>
			)}
			<ButtonBase
				data-testid="cabinet-input-button"
				type="button"
				onClick={handleEditInput}
				className={styles.btn}
			>
				{!isEdited ? <EditIcon /> : <CloseIcon />}
			</ButtonBase>
		</div>
	);
};
