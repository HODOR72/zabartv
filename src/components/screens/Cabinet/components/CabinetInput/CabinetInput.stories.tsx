import { useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { CabinetInput } from './CabinetInput';

export default {
	title: 'CabinetInput',
	component: CabinetInput,
} as ComponentMeta<typeof CabinetInput>;

const DefaultInput: ComponentStory<typeof CabinetInput> = () => {
	const [value, setValue] = useState<string>('aassmoilov@gmail.com');

	return (
		<div style={{ width: '350px' }}>
			<CabinetInput label="Email" placeholder={value} type="email" value={value} applyChanges={setValue} />
		</div>
	);
};

export const Input = DefaultInput.bind({});

const MaskInput: ComponentStory<typeof CabinetInput> = () => {
	const [value, setValue] = useState<string>('11.11.1111');

	return (
		<div style={{ width: '350px' }}>
			<CabinetInput mask='00.00.0000' label="Дата рождения" placeholder={value} value={value} applyChanges={setValue} />
		</div>
	);
};

export const InputWithMask = MaskInput.bind({});
