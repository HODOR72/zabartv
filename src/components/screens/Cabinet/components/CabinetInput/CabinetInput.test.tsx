import { render, screen, fireEvent } from '@testing-library/react';
import { useState } from 'react';
import { CabinetInput } from './CabinetInput';

describe('CabinetInput.tsx', () => {
	const Input = () => {
		const [value, setValue] = useState<string>('aassmoilov@gmail.com');

		return (
			<CabinetInput label="Электронная почта" type="email" value={value} applyChanges={setValue} />
		);
	};

	const MaskInput = () => {
		const [value, setValue] = useState<string>('');

		return (
			<CabinetInput mask='00.00.0000' label="Дата рождения" type="text" value={value} applyChanges={setValue} />
		);
	}

	test('tests', () => {
		render(<Input />);
		const item = screen.getByTestId('cabinet-input');
		const button = screen.getByTestId('cabinet-input-button');
		const input = screen.getByTestId('cabinet-input');

		expect(input).not.toHaveFocus();

		fireEvent.click(button);

		expect(input).toHaveFocus();

		expect(item).toMatchSnapshot();
	});

	test('test input with mask', () => {
		render(<MaskInput />);
		const button = screen.getByTestId('cabinet-input-button');

		const maskInput: HTMLInputElement = screen.getByTestId('cabinet-input-mask');

		expect(maskInput).not.toHaveFocus();

		fireEvent.click(button);

		expect(maskInput).toHaveFocus();

		fireEvent.input(maskInput, {target: {value: '19021998'}})

		expect(maskInput.value).toBe('19.02.1998')
	})
});
