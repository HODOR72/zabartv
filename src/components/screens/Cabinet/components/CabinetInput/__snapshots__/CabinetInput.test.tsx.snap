// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CabinetInput.tsx tests 1`] = `
<input
  class="input"
  data-testid="cabinet-input"
  type="text"
  value=""
/>
`;
